### ISSUE TITLE
Pseudo-random number generation

### ISSUE THREAD

**Issue by blyxyas** [typst/typst#984]
### Description

`calc.random(seed, min, max)` would be a new function under the Calculate category, that takes three optional numbers and, using the first number as a seed, returns a random number between the remaining two numbers.

## To discuss

- Default `min` and `max` values:
    - **Suggested answer**: 0 and 100
- The way to get the random numbers
Rust's standard library doesn't contain a RNG. So we need to pull in a dependency to get those numbers.
    - **Suggested answer**: [`fastrand`](https://crates.io/crates/fastrand)
- Should this function be in `calc` or, should it really be in `math`?

As the function returns an integer, both `min` and `max` would get truncated to be integers.
Seed would be settable.

---

I have a working implementation for this, I'm just opening this issue for discussion before the PR.

### Use Case

Everyone that would need a random number generated in compile-time.

**Comment by PgBiel** [comment]
We've brought this up before (adding random number generators to typst), but there's still the question: would this break memoization? How do you deal with this in your implementation?

**Comment by LaurenzV** [comment]
@PgBiel How would this break memoization? As long as you make it obligatory to provide the seed, the generated numbers would be deterministic and this shouldn't be a problem, or am I missing something?

**Comment by blyxyas** [comment]
Tbh my implementation was made totally without knowing that memoization was used. I'm not currently at home, but when I get there I'll be testing this.

Just a question: Should this be memoizable? I feel like the answer is no.

**Comment by AndyBarcia** [comment]
A random function can be created in typst using mutable state, altought is a bit difficult to use because the value of the random variable can only be acceses as a content value, or through a function.

Here's a very simple example

```
#let s = state("rng", 0)

#let random(min:0, max:100) = {
  s.update(x => x*1103515245+12345)
  
  let dif = max - min
  s.display((x) => calc.mod(calc.abs(x), dif) + min)
}

#random()
#random()
#random()
#random()
#random()

#let random_f(min:0, max:100, f) = {
    s.update(x => x*1103515245+12345)

    let dif = max - min
    s.display((x) => f(calc.mod(calc.abs(x), dif) + min))
}

#let random_circle() = {
    random_f(min: 0, max:100, (x) => {
          place(
              dx: x*1pt,   
              random_f(min: 0, max:100, (y) => {
                  move(
                      dy: y*1pt,   
                      circle(radius: 4pt)
                  )
              })
          )
    })
}

#let n = 0
#while n < 100 {
    random_circle()
    n = n+1
}
```

This give the following results:

![Captura desde 2023-04-26 13-07-48](https://user-images.githubusercontent.com/40731413/234557471-317df3df-450d-49b9-8797-be8122d4a1c8.png)

It might be possible to implement this internally so that the value of the state can be access directly, making it easier to use.

**Comment by blyxyas** [comment]
Maybe using some static values? That would mean [pulling a new dependency](crates.io/crates/lazy_static) or having a static value even for documents that don't use `random()`. Neither are very costly but we should choose.

**Comment by blyxyas** [comment]
> all functions (internal ones included??) must be memoizable.
I don't think this applies to a function that is literally made to return something different every time it's called.

With my implementation the following code:

```
#calc.random()
#calc.random()
#calc.random()
#calc.random()
#calc.random()
#calc.random()
#calc.random()
```

Produces the following output: "34 6 39 0 48 28 5"
If the function uses memoization, the output would be just a group of 34s

**Comment by leo848** [comment]
@blyxyas This is wrong, just as a counter also changes while the syntax and function call is exactly the same. I think something like a pseudo-random number generator that is necessarily seeded might be a good idea, but I strongly dislike the idea of truly non-deterministic outputs. A compiler should ideally be a pure function that takes some inputs and returns some outputs without any external state (although some exceptions may be allowed for convenience, such as reading files during compilation).

**Comment by Enivex** [comment]
> First, I'm not sure why you would want randomness in your document

Not true randomness, but a pseudorandom sequence based on a deterministic seed is useful for such things as generating illustrations or examples. Let's say I'm a teacher, and wanted to generate a whole bunch of problems for my students.

(But that PR isn't the way to do it)

**Comment by PgBiel** [comment]
I think the current date from #435 will work as some sort of seed. We might not have `#now()` for now (which would be even better), but that will work in most cases, probably.

**Comment by blyxyas** [comment]
Then maybe returning a Vec of numbers?

**Comment by Julian-Wollersberger** [comment]
Why is the upper limit an optional parameter? I get that a lower limit of 0 is a good default, but the upper limit is different for most use cases:
* Percentages are usually 0 to 100
* Random element of a list is 0 to length
* Scores are 0 to 5 or o to 10

Also, in other languages you often use modulo like `random() % 256`, which would lead to wildly biased results in typst.

Implementation wise, can `random()` use a document-wide counter internally, and basically output a hash of the counter value? That's random enough for most uses of RNG, and sophisticated RNGs for eg. cryptography just shouldn't be made in typst.

**Comment by laurmaedje** [comment]
> Then maybe returning a Vec of numbers?

That would be more useful I think.

**Comment by jneug** [comment]
Since there was the question for use cases raised: 

I am a teacher and use randomness in LaTeX to generate exercise sheets with random tasks. I have .tex-files to generate basic arithmetic or geometric problems. Whenever I need a new set of problems, I just recompile the file. 

Sometimes we need to generate random passwords for out students, to use for registration at something like a web service. I read the users from a .csv file and generate a password sheet to give out.

So, for my use, a random number generator would be great.

**Comment by laurmaedje** [comment]
I would be fine with adding a PRNG to the stdlib. What I'd rather not do is ship a way to seed it. One possible way to get a seed is to generate a file with a random number with an external program and read it.

**Comment by laurmaedje** [comment]
Although a PRNG might also be better deferred to a package on second thought. There are a bunch of different trade-offs.

**Comment by LaurenzV** [comment]
May I ask why you are against shipping a way to seed it? I would actually think that the best option here is to _force_ the user to provide a seed, so that documents always remain deterministic. What speaks against that in your opinion?

**Comment by Enivex** [comment]
> May I ask why you are against shipping a way to seed it? I would actually think that the best option here is to _force_ the user to provide a seed, so that documents always remain deterministic. What speaks against that in your opinion?

I'm fairly sure what they meant was no way to seed it in a non-deterministic manner.

**Comment by laurmaedje** [comment]
>  I would actually think that the best option here is to force the user to provide a seed, so that documents always remain deterministic.

My wording was a bit unclear, but I indeed meant that you have to provide a seed and that there is no way to get a random seed from Typst.

**Comment by fyuniv** [comment]
> I'm not sure I like this. First, I'm not sure why you would want randomness in your document. And second, in this form, the function does not seem all that useful to me since you basically have to provide one random number (seed) to get one random number. Typically a PRNG turns one seed number into a sequence of random numbers.

Another application is to create different versions of questions for exams. I used to use the LaTeX package `pgfmath` to create questions with random numbers. While I am trying to do the same thing with Typst, I noticed that a random generator is not there yet. It will be great to have such a function.

**Comment by fyuniv** [comment]
> I'm not sure I like this. First, I'm not sure why you would want randomness in your document. And second, in this form, the function does not seem all that useful to me since you basically have to provide one random number (seed) to get one random number. Typically a PRNG turns one seed number into a sequence of random numbers.

Another application is to create different versions of questions for exams. I used to use the LaTeX package `pgfmath`, to create questions with random numbers. While I am trying to do the same thing with Typst, I noticed that a random generator is not there yet. It will be great to have such a function.

**Comment by Roger-luo** [comment]
FYI. see [how jax implements its PRNG](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html) in a immutable fashion, I think something similar is acceptable.

**Comment by Andrew15-5** [comment]
If someone really wants PRNG now, you can use https://github.com/Enter-tainer/jogs (`Math.random()`) or https://github.com/peng1999/typst-pyrunner (`from random import randint, seed; seed(0); randint(0, 100)`).

jogs outputs `float` and pyrunner outputs `str` (convert with `float()` or `int()`).

See https://typst.app/docs/packages/.

**Comment by knuesel** [comment]
There's now the suiji package for pseudo-random number generation: https://typst.app/universe/package/suiji

**Comment by laurmaedje** [comment]
Since this is now well-covered by packages, I'll close this.

**Comment by Enivex** [comment]
> Since this is now well-covered by packages, I'll close this.

The only caveat here is performance. For some purposes you need quite a lot of numbers, and for that I believe suiji is fairly slow. Though I guess that could be solved by offloading to WASM.
